### ISSUE TITLE
torch.argsort descends wrongly

### ISSUE THREAD

**Issue by vergilus** [pytorch/pytorch#15764]
## 🐛 Bug
given a large amount of numbers (say, a list containing 32768 numbers), I found that the argsort behaves wrongly when setting the descend=False.
## To Reproduce
I want to calculate the similarity of word embeddings, and I managed to achieve that by batched-similarity calculation. I expect the most similarity to an embedding should be itself, thus the highest rank should be 1,2... accordingly, and vice versa. 
now say we have torch.tensor object `similarity` (in shape batch_size * 32768), I want to extract top similarities along the last dimension.
```
get_top = torch.argsort(similarity, dim=1)
get_top = torch.argsort(similarity, dim=1, descending=True) 
```
the code gives the result of:

> get_top = torch.argsort(similarity, dim=1)
> [20485, 30807, 27706,  ...,   117,     6,     **1**],
>  [29628, 16835, 23989,  ...,  6773, 10377,     **2**],
>    ...,
>  [30120, 23853, 24914,  ...,   118,    50,    **47**],
>  [23528, 23663, 25283,  ...,    78,    77,    **48**],
>  [31738, 17062, 29731,  ...,  1240,    79,    **49**]

while:

> top8 = torch.argsort(similarity, dim=1, descending=True)
> tensor([[**21820**, 21822, 21816,  ..., 10909, 10910, 32725],
>         [**32497**, 28158, 20713,  ..., 14889, 12632, 16177],
>         [**17110**, 17460, 18861,  ..., 14628, 12113,    18],
>         ...,
>         [**22335**, 21440, 26360,  ..., 14870,  8292, 13842],
>         [**20379**, 16479, 28973,  ..., 11770, 11327, 13454],
>         [**23388**, 19492, 16577,  ...,  7874,  5219, 13230]])
the bold part implys the main problems. and I assume it's not simply because of the accuracy of float numebrs.
## Expected behavior
> top8 = torch.argsort(similarity, dim=1, descending=True)
> tensor([
>         [**1**,  blablabla ],
>         [**2**, blablabla],
>         ...,
>         [**47**, blablabla],
>         [**48**, blablabla],
>         [**49**, blablabla]])

## Environment
 - PyTorch Version: 1.0 stable
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip install
 - Python version: 3.6.3

**Comment by ailzhang** [comment]
Hmmm it's correct when I randomly initiate a tensor. Could you provide more context, or a simple example? 

```
In [12]: a = torch.randn(3, 4)

In [14]: a
Out[14]:
tensor([[ 1.4786,  1.2917,  1.0115,  0.1842],
        [-0.5320,  0.2807, -0.0722,  0.6151],
        [ 1.3300, -0.6290,  0.0619,  0.3336]])

In [16]: torch.argsort(a, dim=1)
Out[16]:
tensor([[3, 2, 1, 0],
        [0, 2, 1, 3],
        [1, 2, 3, 0]])

In [17]: torch.argsort(a, dim=1, descending=True)
Out[17]:
tensor([[0, 1, 2, 3],
        [3, 1, 2, 0],
        [0, 3, 2, 1]])
```

**Comment by fmassa** [comment]
Does your tensor contain repeated entries? In yes, then that's probably the problem: I believe we don't guarantee which order repeated entries will get output in argsort

**Comment by vergilus** [comment]
@ailzhang  I think you should try much larger amount of entries. for example, you can try randn[3, 32768]. 
actually I presume that topk values should be the same, but its not.

**Comment by vergilus** [comment]
@fmassa no. btw, the multiple-entries thing does not explain why the top similarity index of a vec is not itself. 
I tried torch.tensor.topk(), it yields topk results just as the argsort() 's last k results (in the opposite order of course). It seems to happen when you try to deal with large amounts of entries.

**Comment by vergilus** [comment]
@ailzhang 
[log.zip](https://github.com/pytorch/pytorch/files/2732960/log.zip)

            a = numpy.load("log")
            print(a)
            a=torch.tensor(a)
            print (a.topk(top_reserve, dim=0)[1])
            print (torch.argsort(a, dim=0, descending=True))

**Comment by ailzhang** [comment]
Okay I think the problem is caused by `NaN`. Here's a simpler repro:
```
In [33]: a = torch.tensor([3, float('NaN'), 2])

In [34]: torch.argsort(a, dim=0)
Out[34]: tensor([0, 1, 2]) # wrong!

In [36]: a = torch.tensor([1, float('NaN'), 2])

In [37]: torch.argsort(a, dim=0, descending=True)
Out[37]: tensor([0, 1, 2]) # wrong!
```
Similarly if we remove the leading `Nan` in the array you provided, the problem is gone.  
```
In [38]: a = numpy.load("log")

In [39]: a = torch.tensor(a)

In [40]: b = a[1:]

In [41]: b
Out[41]: tensor([ 0.0791, -0.0522,  0.0226,  ..., -0.0267,  0.1027,  0.0064])

In [49]: print(torch.argsort(b, dim=0, descending=True).narrow(0,0,8))
tensor([ 49, 117,  46,  12, 143,  16,  19, 327])

In [50]: print(torch.argsort(b, dim=0).narrow(0,b.shape[0]-8,8))
tensor([327,  19,  16, 143,  12,  46, 117,  49])
```
Actually the same problem exists for `torch.sort` as well, that `nan` breaks the monotonic increasing in the middle, but it was hidden in the printed sequence. 
```
In [51]: torch.sort(a)
Out[51]:
(tensor([-0.2427, -0.2274, -0.2272,  ...,  0.5829,  0.7206,  1.0000]),
 tensor([26606, 25901, 32186,  ...,    47,   118,    50]))
In [57]: torch.sort(a)[0][16360:16365]
Out[57]: tensor([ 0.2460,  0.2769,     nan, -0.2238, -0.2159])
```

**Comment by ailzhang** [comment]
FYI numpy handles this case by always putting `nan` at the end the sorted sequence.  We should probably follow numpy behavior. 

```
In [14]: a
Out[14]:
array([        nan,         inf,        -inf, ..., -0.02669582,
        0.10268118,  0.0064341 ], dtype=float32)

In [15]: numpy.sort(a)
Out[15]:
array([       -inf, -0.24265206, -0.22741349, ...,  1.        ,
               inf,         nan], dtype=float32)

In [16]: numpy.argsort(a)
Out[16]: array([    2, 26606, 25901, ...,    50,     1,     0])
```

**Comment by umanwizard** [comment]
Taking a look

**Comment by umanwizard** [comment]
FWIW this matches vanilla python behavior:
```
>>> sorted([3, float('NaN'), 2])
[3, nan, 2]
```
